Fix LoRA indices for interleaved QKV weights

ebsmothers opened the pull request:
As discussed in #891, the Q, K, and V projections are interleaved in Llama2 checkpoints and so the `zero_pad` method of `LoRAQKVLinear` is not being correctly applied. This PR modifies `self.lora_ind` accordingly to handle this case.

Test plan:
- Updated the expected LoRA indices test_lora.py's `test_lora_mqa_gqa` as per @Andrei-Aksionov's suggestion.
- Added extra tests in `test_lora_mqa_gqa` where we set the layer's linear weights to 0 (so only outputs with LoRA applied should be nonzero), then confirm that all the interleaved Q and V indices are nonzero, while K indices are all zero.
```
python -m pytest  -v tests/test_lora.py
...
========= 189 passed, 2 skipped, 10 warnings in 11.81s ================
```

Fixes #891

Andrei-Aksionov commented on lit_gpt/lora.py:
```diff
@@ -271,13 +271,42 @@ def __init__(
+
+            q_per_kv = self.n_head // self.n_query_groups
+            slice_size = q_per_kv + 2
+            head_dim = out_features // (num_kv_heads * slice_size)
```
```suggestion
            head_dim = out_features // (self.n_query_groups * slice_size)
```

Andrei-Aksionov commented on lit_gpt/lora.py:
```diff
@@ -271,13 +271,42 @@ def __init__(
+            # times the size of each k and v block.
+
+            q_per_kv = self.n_head // self.n_query_groups
+            slice_size = q_per_kv + 2
```
I vote for changing `slice_size` to `query_group_size`.

Andrei-Aksionov commented on lit_gpt/lora.py:
```diff
@@ -271,13 +271,42 @@ def __init__(
             if enable_v:
-                self.lora_ind.extend(range(self.linear.in_features + self.kv_embd_size, self.linear.out_features))
+                v_ind = list(filter(lambda x: (x // head_dim) % slice_size == slice_size - 1, ind))
+                self.lora_ind.extend(v_ind)
```
Final indices must be in a sorted order.
```python
self.lora_ind.sort()
```

The reason being is that in the `zero_pad` method we use `torch.Tensor.index_copy` and, according to the [docs](https://pytorch.org/docs/stable/generated/torch.Tensor.index_copy_.html), the values will be copied in the exact order that is specified by provided indices.
In other words, when we want to pad a matrix, which structure is (q, q, q, q, k, v, ...) and first 6+ indices are for query matrix, then k and v will be copied to where queries are stored.

carmocca replied:
I didn't understand this. Can you show your point in a bit of code?

Andrei-Aksionov replied:
Sure, but it turns out that it's not necessary 🙂.
As I understand, there is no need for LoRA weights to repeat the same interleaved placement as in the pretrained weights. If we do not sort indices, the final LoRA weights will have structure [Q1, Q2, K1, K2, V1, V2] (if it's applied to all layers) and the correct placement will be done via `zero_pad` method.

Here are the code examples (LoRA is applied to Q and V matrices).

1. If we don't sort indices and LoRA has no interleaved placement of weights - everything is fine.

```python
import torch

n_embd = 8
n_head = 4

padded = torch.zeros((n_embd * 3), dtype=int)
print(f"{padded=}")
>> padded=tensor([0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0, 0])

# NOTE: 1 - query updates, 3 - value updates
lora = torch.tensor([1, 1] * n_head + [3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 1, 1, 1, 1, 1, 1, 3, 3, 3, 3, 3, 3, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3])
```

2. If the indices are unsorted, but LoRA weights are interleaved, then the final order isn't correct.

```python
lora = torch.tensor([1, 1, 3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 1, 1, 3, 3, 0, 0, 3, 3, 1, 1, 0, 0, 1, 1, 3, 3, 0, 0, 3, 3])
```

3, With sorted indices and interleaved LoRA weights, everything is fine.

```python
lora = torch.tensor([1, 1, 3, 3] * n_head)
print(f"{lora=}")
>> lora=tensor([1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3, 1, 1, 3, 3])

indices = [0, 1, 6, 7, 12, 13, 18, 19, 4, 5, 10, 11, 16, 17, 22, 23]
indices.sort()
indices = torch.tensor(indices)

padded.index_copy(0, indices, lora)
>> tensor([1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3, 1, 1, 0, 0, 3, 3])
```

----

I believe that technically all three cases might work, though I think that in order to learn, the model will converge to either the first or the third case. Since the third case requires sorting, and it's unnecessary, I think I have to admit that **I was wrong and there is no need to sort indices**.

ebsmothers replied:
Thanks for the detailed analysis, in this case I will go with option (1) to avoid the sort.

Andrei-Aksionov commented:
@ebsmothers Thanks for the PR 👍 
I've left a couple of comments.
In addition, [this](https://github.com/Lightning-AI/lit-gpt/blob/93b0dc75158f287520db016a6e21f71de865104f/tests/test_lora.py#L82) test needs to be updated to reflect the interleaved placement in `self.lora_ind`.

Other than that, looks good.
Thanks a lot.

carmocca reviewed (commented):
Do you recommend that this is tested numerically? For instance loading fixed pretrained weights and asserting the output of `LoRAQKVLinear.forward`

We will run an ablation with real training scripts before merging too

Andrei-Aksionov commented:
> Do you recommend that this is tested numerically? For instance loading fixed pretrained weights and asserting the output of `LoRAQKVLinear.forward`

Yep, something like this. 
1. create pretrained weights with zero values
2. run a forward method
3. assert that `lora_ind` are correct
4. assert that values under `lora_ind` aren't zero
5. assert that values that aren't under `lora_ind` are zero

ebsmothers commented:
Thanks for the comments, I've updated everything and have tests passing so will formally open this up for review now. Please let me know if there's anything else you need from my end prior to running the ablation.

Btw the zero_pad docstring may also need an update, which I haven't touched yet. Happy to copy what I added to \_\_init__.py or leave it as is in this PR, let me know what you prefer.

Andrei-Aksionov commented:
Thanks @ebsmothers for making time to finish the work.

> Btw the zero_pad docstring may also need an update, which I haven't touched yet. Happy to copy what I added to __init__.py or leave it as is in this PR, let me know what you prefer.

I think it's better to put your updated docstring in the `zero_pad` method. In the init method just keep:
```text
# Compute the indices
# Indices are needed to properly pad the weight updates with zeros in `zero_pad` method.
```

carmocca commented on lit_gpt/lora.py:
```diff
@@ -261,23 +261,21 @@ def __init__(
-            # ----------------------------------------
+            # Indices are needed to properly pad weight updates with zeros in `zero_pad` method.
+            q_per_kv = self.n_head // self.n_query_groups
+            query_group_size = q_per_kv + 2
```
We should use the same name here as in the model: https://github.com/Lightning-AI/lit-gpt/blob/main/lit_gpt/model.py#L197C9-L197C18

```suggestion
            total_qkv = q_per_kv + 2
```

carmocca commented on lit_gpt/lora.py:
```diff
@@ -261,23 +261,21 @@ def __init__(
+            # Indices are needed to properly pad weight updates with zeros in `zero_pad` method.
+            q_per_kv = self.n_head // self.n_query_groups
+            query_group_size = q_per_kv + 2
+            head_dim = out_features // (self.n_query_groups * query_group_size)
```
Similarly

```suggestion
            head_size = out_features // (self.n_query_groups * query_group_size)
```

carmocca commented on lit_gpt/lora.py:
```diff
@@ -261,23 +261,21 @@ def __init__(
-                self.lora_ind.extend(range(0, self.linear.in_features))
+                q_ind = list(filter(lambda x: (x // head_dim) % query_group_size < query_group_size - 2, ind))
+                self.lora_ind.extend(q_ind)
             if enable_k:
-                self.lora_ind.extend(range(self.linear.in_features, self.linear.in_features + self.kv_embd_size))
+                k_ind = list(filter(lambda x: (x // head_dim) % query_group_size == query_group_size - 2, ind))
+                self.lora_ind.extend(k_ind)
             if enable_v:
-                self.lora_ind.extend(range(self.linear.in_features + self.kv_embd_size, self.linear.out_features))
+                v_ind = list(filter(lambda x: (x // head_dim) % query_group_size == query_group_size - 1, ind))
```
nit: Can you change this to a list comprehension instead of list+filter+lambda? It's generally easier to grasp for beginners

carmocca commented on lit_gpt/lora.py:
```diff
@@ -293,6 +291,27 @@ def zero_pad(self, x: torch.Tensor) -> torch.Tensor:
+
+        [Q, Q, Q, Q, K, V, Q, Q, Q, Q, K, V, Q, Q, Q, Q, K, V],
+
+        where each Q, K, and V has size head_dim.
```
```suggestion
        where each Q, K, and V has size head_size.
```

carmocca reviewed (approved):
Some nits.

@awaelchli Would you be so kind to try out this branch?

awaelchli commented:
I compared running LoRA finetuning before and after this PR. I took the code from [this Studio](https://lightning.ai/lightning-ai/studios/instruction-finetuning-tinyllama-1-1b-llm?view=public&section=recent) and only checked out the relevant commits for the comparison. 

The LoRA settings were:
```
lora_r = 8
lora_alpha = 16
lora_dropout = 0.05
lora_query = True
lora_key = False
lora_value = True
lora_projection = False
lora_mlp = False
lora_head = False
```

I'm not able to see a noticeable impact on the optimization or performance of the model. An explanation for this might be that SGD is happy to adjust to the garbled indices over the interleaved matrices so it doesn't impact optimization. Let me know if you need to see any additional results @carmocca. Changes LGTM

ebsmothers commented:
Sorry for going MIA the past few days. Thanks so much @carmocca for taking this across the finish line!
